Lib/fontTools/ttLib/tables/TupleVariation.py

```diff
@@ -446,6 +444,125 @@ def getTupleSize_(flags, axisCount):
 			size += axisCount * 4
 		return size
 
+	def getDeltaType(self):
+		""" Check if deltas are (x, y) as in gvar, or single values as in cvar.
+		Returns a string ("gvar" or "cvar"), or None if empty.
+		"""
+		firstDelta = next((c for c in self.coordinates if c is not None), None)
+		if firstDelta is None:
+			return  # empty or has no impact
+		if type(firstDelta) is tuple and len(firstDelta) == 2:
+			return "gvar"
+		elif type(firstDelta) in (int, float):
+			return "cvar"
+		else:
+			raise TypeError("invalid type of delta: %s" % type(firstDelta))
+
+	def scaleDeltas(self, scalar):
+		if scalar == 1.0:
+			return  # no change
+		deltaType = self.getDeltaType()
```

Comment:
Looks repetitive and self-similar.

if you got the tuple width here could you generalize the code? E.g. if scaler 0 self.coordinates = [tuple of <width> 0's] * len(self.coordinates). Then we just need to collapse the else branches :)

Reply:
sorry, I didn't get this. WDYM by "tuple width"? The value of self.coordinates is either a tuple (as in the python built-in type) or a float.

Reply:
(it bugs me we have all the branches with almost identical code on both branches)

Reply:
TupleVariation.coordinates is either a list of tuples of floats ("gvar"), or a list of floats ("cvar"). I cannot collapse the else branches in this case. I prefer not to change the TupleVariation class too much.

Reply:
Consider the branches to multiply by 0. You are just trying to multiply by 0 or by (0,0). So, pseudo-code:

```
width = getDeltaWidth()
if scalar == 0:
  self.coordinates = _coordMultiplier(width, 0) # returns 0 for width 1, (0, 0) for width 2
```

But ... why branch on scalar 0 at all? - multiplying by 0 often gets you to 0. So, replace L465..480 with something like:

```
_scaleCoordinates(scalar)

def scaleCoordinates(scalar):
  coord_width = getCoordWidth()
  multiplier = scalar
  if coord_width > 1:
    multiplier = tuple([scalar] * coord_width)
  # ...now multiply all the coordinates by multiplier...
  # if you really care that for scalar 0 that None => 0 you can code that here too
```

Reply:
> You are just trying to multiply by 0 or by (0,0)

i'm confused. I'm not "multiplying" by 0 or (0, 0). I'm using the `*` operator on a list object together with an int operand in order to obtain a new list of given length, with the content of the original list repeated.
One cannot just multiply tuples together. I want to multiply their content (floats) by another float (scalar).
The code is complicated by the fact that TupleVariation.coordinates is a `List[Optional[Union[Tuple[float, float], float]]]`, e.g.: these are all valid values: `[None, None]`, `[(0, 0), (1, 1), None, (3, 3)]`, or `[1, 2, None, 4, 5]`.

Reply:
My suggestion is essentially implement a general purpose multiply coords instead of all this branching.

Looking at 0s specifically it appears to me the code that says if scalar = 0 coordinates = [0] * # of coords is just short-circuit for the notion that if we multiplied each coord by 0 we'd get all 0's. And then we code it all over again to support the case where we have two coordinates. Then repeat for non-zero multiplication once for each width.

Reply:
I removed the special case for scalar = 0, but we still need to branch out when scaling, adding, rounding, etc. depending on whether the coordinate is a bare number or a tuple of numbers.